### PR TITLE
Better error array numerics

### PR DESCRIPTION
**Pull request by xunilrj**
## Description

This PR fixes: https://github.com/FuelLabs/sway/issues/6376

The issue is that we use the first element of the array as its element type. This can be problematic as the first element is not "concrete". For example, when it is `Numeric`.

Now we iterate the array elements, search for the first "good" element, and use its type as the array type. This means:
1 - if the type is concrete, we use it;
2 - if the type is `Numeric`, we keep searching for something better;

This allows consistent arrays, but where the concrete types are at the end.

```sway
let a = [1, 2u8]; <- we will choose u8
let a = [None, Some(1), Some(2u8)]; <- we will choose Option<u8>, instead of Option<T>
```

## snapshot tests

This PR also brings an experiment. Now when an "e2e"  test has a file `snapshot.toml` it will run as `cargo insta` snapshot tests. These tests can be run as normal: `cargo r -p test`, and in two new ways:

```
> cargo t -p test
> cargo insta test
```

Snapshots can be reviewed using normal "cargo insta" workflow.

The secret lies in https://github.com/LukasKalbertodt/libtest-mimic. **I also want to suggest that all our tests migrate to this model**, where we could just use "cargo test" normally. Apart from that, I opted to use the `forc` executable, which means tests will run in parallel. In the future, if we want to poke compiler internals, we can expose them using flags, or create a special binary to fit our needs.

`snapshot.toml` is still empty because I have no idea to what put there at the moment. My initial idea was to configure what ends up into snapshots (multiple runs, compiler flags, script/contract calls) etc... But for the moment, empty is enough.

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
   - [ ] If my change requires substantial documentation changes, I have [requested support from the DevRel team](https://github.com/FuelLabs/devrel-requests/issues/new/choose)
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Comment by Voxelot** [comment]
+1 for libtest mimic, we've been using it over in the fuel-core repo already as well.

**Review by ironcev — COMMENTED** [review]
Great that we are starting with snapshot tests!! :100: :1st_place_medal:  Can we just add to the README.md of the `e2e_vm_tests` short instructions how to use them? Basically this what is written in the PR description.

**Comment by xunilrj** [comment]
> Great that we are starting with snapshot tests!! 💯 🥇 Can we just add to the README.md of the `e2e_vm_tests` short instructions how to use them? Basically this what is written in the PR description.

Done.

**Comment by xunilrj** [comment]
> This means we lose the way to see individual test outputs when running the test suite ([LukasKalbertodt/libtest-mimic#9](https://github.com/LukasKalbertodt/libtest-mimic/issues/9)).

We can still use what we have: `gag` would still work, but ideally, we would move to a better model, allowing us to run more tests in parallel.

I think that for compilation, we can use the `forc` executable like I am doing for snapshots. `cargo insta` would be much better for warnings/errors for both `should_pass` and `should_fail` tests.

For tests that need the VM to run, we can keep what we have because we do not capture output, right?

For "IR generation tests", we would need to change them to start using `forc` and use `cargo insta` with `--ir initial`.

"sdk harness" can also benefit from this workflow, but we would need to build the `sway` code using `build.rs`, because of all the macros that need access to the ABI jsons. So I would probably keep them in a different binary and only compile them when needed.

> Another potential issue is the filtering support we have built in our test runner over time, which we would need to reimplement on top of the new system.

I think we have two options. One is to keep the current filter and only return the filtered tests to `libmimic_test`. The only difference I would propose is that when no filter is specified, we would expand all tests and all options. For example, no filter runs both debug **AND** release etc...

A second option would be to always expand everything single option, and let `cargo t` filter do its thing. So for example, instead of

```
> cargo r -p test -- --abi-only
```

It could be something like

```
> cargo r -p test -- "(abi)"
```

This would work because our "test finder" can append "(abi)"  when a test, tests the abi. I think we can think of something similar to other cases.

> What are we gaining by introducing libtest mimic? Is it just to make the output as similiar to `cargo test`? Or is there anything else?

I think the biggest gain would be that our tests would be more "standard rust tests", which means they would work with the rest of the ecosystem. `cargo insta`, `cargo nextest` are two examples.

Maybe even things like: https://github.com/swellaby/vscode-rust-test-adapter

But, for me, honestly, would be that `cargo t` actually tests everything.

**Review by IGI-111 — DISMISSED** [review]

File: sway-core/src/type_system/id.rs [hunk 1]
```diff
@@ -27,7 +27,7 @@ pub enum IncludeSelf {
     No,
 }
 
-pub enum IncludeNumeric {
+pub enum NumericIsNonConcrete {
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply sway-core/src/type_system/id.rs:30]: But `Yes` means that `Numeric` is not concrete.

File: test/tests/tests.rs [hunk 2]
```diff
@@ -0,0 +1,122 @@
+use libtest_mimic::{Arguments, Trial};
+use std::{path::PathBuf, sync::Once};
+
+static FORC_COMPILATION: Once = Once::new();
+
+fn compile_forc() {
+    // e2e_tests::test_consistency::check().unwrap();
```
[the remark this answers is not in the document]
Comment (xunilrj) [reply test/tests/tests.rs]: Fixed.

File: sway-core/src/type_system/id.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (IGI-111) [reply sway-core/src/type_system/id.rs:30]: Can we avoid double negation and have `NumericState::{Concrete, Abstract}` or something?

[the remark this answers is not in the document]
Comment (xunilrj) [reply sway-core/src/type_system/id.rs:30]: I was avoiding changes to the other usages. But I can try something.

**Review by IGI-111 — APPROVED** [review]

**Pull request merged**